cm_service_manager.py

```diff
@@ -0,0 +1,100 @@
+parser.add_argument('-l', '--cluster', required=False,
+                    help="cluster to restart service in ('mycluster')")
+parser.add_argument('-s', '--service', required=False,
+                    help="service to restart ('hdfs')")
+
+args = parser.parse_args()
+print args
+
+# https://cloudera.github.io/cm_api/docs/quick-start/
+api = ApiResource(args.cm_host, username=args.username, password=args.password)
+
+def select_from_items(item_type, items):
+    print("\n".join(["Choose a %s" % item_type] + [i.lower() for i in items]))
+    item = raw_input().lower()
+    print "Input received: %s" % item
+    if item in items:
+        return item
+    else:
+        print 'Bad input received...'
+        return select_from_items(item_type, items)
+
+
+# Rolling restart method (using few defaults - Role types, role names. It will take all)
+def rolling_restart(service_rr):
+    print service_rr
+    batch_size = raw_input("Number of worker roles to restart together in a batch")
+    fail_count_threshold = raw_input("Stop rolling restart if more than this number of worker batches fail")
+
+    options = ["true", "false"]    # options for stale config
+    stale_config_only = select_from_items('option to Restart roles on nodes with stale config? Default is false', options)   # Needs to be a user selection
+
+    # rr_command = service_rr.rolling_restart(batch_size, fail_count_threshold, None, stale_config_only, None, None, None).wait()
+    # return rr_command
+
+# Get the cluster
+clusters = api.get_all_clusters()
+cluster_names = [c.name.lower() for c in clusters]
+cluster_name = ''
+if args.cluster is not None:
+    cluster_name = args.cluster.lower()
+else:
+    cluster_name = select_from_items('cluster', cluster_names)
+
+for cluster in clusters:
+    if cluster.name.lower() == cluster_name:
+
+        # Get the service
+        services = cluster.get_all_services()
+        service_names = [s.name.lower() for s in services]
+        service_name = ''
+        if args.service is not None:
+            service_name = args.service.lower()
+        else:
+            service_name = select_from_items('service', service_names)
+
+        # Get the action to be performed on predefined actions
+        defined_actions = ['Service restart', 'Deploy Client config', 'Rolling Restart']
+
+        for service in services:
+            if service.name.lower() == service_name:
+
+                action = select_from_items('action', [i.lower() for i in defined_actions])
+
+                if action == defined_actions[0].lower():  # Restart the service
+                    command = service.restart().wait()
+
+                elif action == defined_actions[1].lower():    # Deploy client configs
+                    command = service.deploy_client_config().wait()
+
+                elif action == defined_actions[2].lower() and ('hue' in service.name.lower() or 'hive' in service.name.lower() or \
```

Comment:
I'd still rather just call it and then catch the exception - its more pythonic and it handles the services that you havent listed here. you shouldnt be listing out all these services like this.

you can even make it so that if rolling restart is selected and not available then you just do a restart